File: Nori/Store/HistorySearch.swift
import Foundation

/// Search over the history: case- and diacritic-insensitive substring first,
/// then an in-order subsequence match for typo-tolerant "fuzzy" hits.
enum HistorySearch {
    struct Match: Equatable {
        /// Lower is better. Substring hits score below fuzzy hits.
        var score: Double
        /// Ranges in `title` to highlight (may be empty when the hit is only in the full text).
        var titleRanges: [Range<String.Index>]
    }

    static func match(query: String, title: String, searchText: String) -> Match? {
        let needle = query.trimmingCharacters(in: .whitespaces)
        guard !needle.isEmpty else { return Match(score: 0, titleRanges: []) }
        let options: String.CompareOptions = [.caseInsensitive, .diacriticInsensitive]

        // 1. Every whitespace-separated term must appear somewhere (in title or full text).
        let terms = needle.split(whereSeparator: \.isWhitespace).map(String.init)
        var ranges: [Range<String.Index>] = []
        var allTermsFound = true
        var bonus = 0.0
        for term in terms {
            if let range = title.range(of: term, options: options) {
                ranges.append(range)
                if range.lowerBound == title.startIndex { bonus -= 0.1 }
            } else if searchText.range(of: term, options: options) != nil {
                bonus += 0.2
            } else {
                allTermsFound = false
                break
            }
        }
        if allTermsFound {
            return Match(score: 1 + bonus + Double(title.count) / 100_000, titleRanges: mergeRanges(ranges))
        }

        // 2. Fuzzy: characters of the query appear in order in the title.
        guard needle.count >= 2 && needle.count <= 32 else { return nil }
        return subsequenceMatch(needle: needle, haystack: String(title.prefix(300)))
    }

    private static func subsequenceMatch(needle: String, haystack: String) -> Match? {
        let needleChars = Array(needle.lowercased())
        let haystackLower = haystack.lowercased()
        var ranges: [Range<String.Index>] = []
        var needleIndex = 0
        var gapPenalty = 0.0
        var lastMatchOffset = -1
        var offset = 0
        var index = haystackLower.startIndex
        while index < haystackLower.endIndex, needleIndex < needleChars.count {
            if haystackLower[index] == needleChars[needleIndex] {
                let next = haystackLower.index(after: index)
                ranges.append(index..<next)
                if lastMatchOffset >= 0 { gapPenalty += Double(offset - lastMatchOffset - 1) }
                lastMatchOffset = offset
                needleIndex += 1
            }
            index = haystackLower.index(after: index)
            offset += 1
        }
        guard needleIndex == needleChars.count else { return nil }
        // Convert to indices of the original string (lowercasing preserves counts for our purposes).
        let originalRanges = ranges.compactMap { range -> Range<String.Index>? in
            let start = haystackLower.distance(from: haystackLower.startIndex, to: range.lowerBound)
            guard let lower = haystack.index(haystack.startIndex, offsetBy: start, limitedBy: haystack.endIndex),
                  lower < haystack.endIndex else { return nil }
            return lower..<haystack.index(after: lower)
        }
        let score = 5 + gapPenalty / Double(max(haystack.count, 1)) * 10 + Double(gapPenalty) * 0.01
        return Match(score: score, titleRanges: mergeRanges(originalRanges))
    }

    static func mergeRanges(_ ranges: [Range<String.Index>]) -> [Range<String.Index>] {
        let sorted = ranges.sorted { $0.lowerBound < $1.lowerBound }
        var merged: [Range<String.Index>] = []
        for range in sorted {
            if let last = merged.last, range.lowerBound <= last.upperBound {
                merged[merged.count - 1] = last.lowerBound..<max(last.upperBound, range.upperBound)
            } else {
                merged.append(range)
            }
        }
        return merged
    }
}

File: windows/Nori.Core/HistorySearch.cs
using System.Globalization;

namespace Nori.Core;

/// <summary>A highlighted run in a title: UTF-16 offset and length.</summary>
public readonly record struct TextRange(int Start, int Length)
{
    public int End => Start + Length;
}

/// <summary>
/// Search over the history: case- and diacritic-insensitive substring first, then an in-order
/// subsequence match for typo-tolerant "fuzzy" hits.
/// </summary>
public static class HistorySearch
{
    public sealed record Match(double Score, IReadOnlyList<TextRange> TitleRanges);

    private static readonly CompareInfo Compare = CultureInfo.InvariantCulture.CompareInfo;
    private const CompareOptions Options = CompareOptions.IgnoreCase | CompareOptions.IgnoreNonSpace | CompareOptions.IgnoreKanaType | CompareOptions.IgnoreWidth;

    public static Match? Find(string query, string title, string searchText)
    {
        var needle = query.Trim();
        if (needle.Length == 0) return new Match(0, []);

        // 1. Every whitespace-separated term must appear somewhere (in title or full text).
        var terms = needle.Split((char[]?)null, StringSplitOptions.RemoveEmptyEntries);
        var ranges = new List<TextRange>();
        var allTermsFound = true;
        var bonus = 0.0;
        foreach (var term in terms)
        {
            if (IndexOf(title, term) is { } range)
            {
                ranges.Add(range);
                if (range.Start == 0) bonus -= 0.1;
            }
            else if (IndexOf(searchText, term) is not null)
            {
                bonus += 0.2;
            }
            else
            {
                allTermsFound = false;
                break;
            }
        }
        if (allTermsFound)
        {
            return new Match(1 + bonus + title.Length / 100_000.0, MergeRanges(ranges));
        }

        // 2. Fuzzy: characters of the query appear in order in the title.
        if (needle.Length is < 2 or > 32) return null;
        return SubsequenceMatch(needle, title.Length > 300 ? title[..300] : title);
    }

    /// <summary>Culture-insensitive, accent-insensitive substring search that reports the matched run.</summary>
    public static TextRange? IndexOf(string haystack, string needle)
    {
        if (haystack.Length == 0 || needle.Length == 0) return null;
        var index = Compare.IndexOf(haystack.AsSpan(), needle.AsSpan(), Options, out var length);
        if (index < 0) return null;
        return new TextRange(index, Math.Max(length, 1));
    }

    private static Match? SubsequenceMatch(string needle, string haystack)
    {
        var needleChars = needle.Select(char.ToLowerInvariant).ToArray();
        var ranges = new List<TextRange>();
        var needleIndex = 0;
        var gapPenalty = 0.0;
        var lastMatchOffset = -1;
        for (var offset = 0; offset < haystack.Length && needleIndex < needleChars.Length; offset++)
        {
            if (char.ToLowerInvariant(haystack[offset]) != needleChars[needleIndex]) continue;
            ranges.Add(new TextRange(offset, 1));
            if (lastMatchOffset >= 0) gapPenalty += offset - lastMatchOffset - 1;
            lastMatchOffset = offset;
            needleIndex++;
        }
        if (needleIndex != needleChars.Length) return null;
        var score = 5 + gapPenalty / Math.Max(haystack.Length, 1) * 10 + gapPenalty * 0.01;
        return new Match(score, MergeRanges(ranges));
    }

    public static IReadOnlyList<TextRange> MergeRanges(IEnumerable<TextRange> ranges)
    {
        var merged = new List<TextRange>();
        foreach (var range in ranges.OrderBy(r => r.Start))
        {
            if (merged.Count > 0 && range.Start <= merged[^1].End)
            {
                var last = merged[^1];
                merged[^1] = new TextRange(last.Start, Math.Max(last.End, range.End) - last.Start);
            }
            else
            {
                merged.Add(range);
            }
        }
        return merged;
    }
}
